### DISCUSSION TITLE
Pass custom parameters to mutator

### DISCUSSION THREAD

**Discussion by ghost** [orval-labs/orval#1610] [Q&A] [answered]
In the mutator function I need more information available to transform the data I receive.

Is there a way to pass the definition `$ref` to the mutator (or something like that)?

**Comment by scottdickerson** [comment]
I have a similar usecase for both mutators and GET API calls, I need to extend the generated API calls to require a `customParam` parameter like this: 

```
export const getData = async (
  params?: GetDataParams,
  options?: RequestInit,
  customParam: string,
): Promise<getDataResponse> => {
  return fetchWrapper<Promise<getDataResponse>>(
    getGetUpcomingAuctionsUrl(params),
    {
      ...options,
      method: 'GET',
    },
    customParam
  );
};
```

Is there anyway to add custom parameters to the generated API functions.

My current orval config is this
```
 input: './specs/api/openapi.yaml',
    output: {
      mode: 'split',
      schemas: './generated/api/types',
      target: './generated/api/api.ts',
      mock: { type: 'msw', delay: 10, useExamples: false },
      prettier: true,
      client: 'fetch',
      override: {
        mutator: {
          path: './api/fetchWrapper.ts', 
          name: 'fetchWrapper',
        },
      },
    },
```

**Reply by ghost** [reply, marked as answer]
We have accomplished this using `jscodeshift` ([GitHub](https://github.com/facebook/jscodeshift)).

You pass all files to your `jscodeshift` transform (e.g. `transform.cjs`) using `orval`s `afterAllFilesWrite` setting:

```js
import {run} from 'jscodeshift/src/Runner.js';
// snip
afterAllFilesWrite: [paths => run('./transform.cjs', paths, {parser: 'ts'})]
```

And in your `transform.cjs` you find the callee ending with `Mutator` and you add another argument.

It is kind of rough, but it works and it is not too much additional source code.
